reflect/src/main/java/dagger/reflect/BindingMap.java

```diff
@@ -80,9 +85,37 @@ Builder add(Key key, Binding binding) {
       return this;
     }
 
+    Builder addIntoSet(Key key, Binding binding) {
+      if (key == null) throw new NullPointerException("key == null");
+      if (binding == null) throw new NullPointerException("binding == null");
+
+      List<Binding> bindings = setBindings.get(key);
+      if (bindings == null) {
+        bindings = new ArrayList<>();
+        setBindings.put(key, bindings);
+      }
+      bindings.add(binding);
+      return this;
+    }
+
     BindingMap build() {
-      // Take a defensive copy in case the builder is being used to create multiple instances.
       ConcurrentHashMap<Key, Binding> allBindings = new ConcurrentHashMap<>(bindings);
+
+      // Coalesce all of the bindings for each key into a single set binding.
+      for (Map.Entry<Key, List<Binding>> setBinding : setBindings.entrySet()) {
+        Key elementKey = setBinding.getKey();
+        Key setKey = Key.of(elementKey.qualifier(),
+            new ParameterizedTypeImpl(null, Set.class, elementKey.type()));
+
+        // Take a defensive copy in case the builder is being re-used.
+        List<Binding> value = new ArrayList<>(setBinding.getValue());
```

Comment:
Whoops. This should probably get a better name. Probably the default from extract variable intention action.